### ISSUE TITLE
Issue: VAE Encoding in FoleyCrafter Cannot Reconstruct Mel Spectrogram

### ISSUE THREAD

**Issue by ISSpeechSEC** [open-mmlab/FoleyCrafter#29]
Hi, I encountered an issue when using the pre-trained VAE model in FoleyCrafter.

When I encode a Mel spectrogram into latents and then decode it back, the reconstructed spectrogram does not match the original, and the vocoder outputs noise. However, in the inference code, the decoding process alone works well.

I suspect that there may be additional image processing applied to the spectrogram before encoding. Could you clarify if there are any preprocessing steps required before encoding?

Here is the code I used:

import os
import argparse
import torch
import torchaudio
import soundfile as sf
from tqdm import tqdm
from foleycrafter.pipelines.auffusion_pipeline import Generator
import librosa
from foleycrafter.utils.converter import (
    dynamic_range_compression_torch,
    MAX_WAV_VALUE,
    mel_spectrogram  # 关键修复：导入缺失的函数
)
from diffusers.models import AutoencoderKL, ControlNetModel, ImageProjection
from foleycrafter.pipelines.auffusion_pipeline import Generator, denormalize_spectrogram
from diffusers.image_processor import PipelineImageInput, VaeImageProcessor

def args_parse():
    config = argparse.ArgumentParser()
    config.add_argument("--wav_dir", type=str, default="/data/llm_data/FoleyCrafter2/FoleyCrafter-main/LLM_Exp/small_dataset_wav")
    config.add_argument("--reconstructed_wav_dir", type=str, default="/data/llm_data/FoleyCrafter2/FoleyCrafter-main/LLM_Exp/reconstructed_wav")
    config.add_argument("--vocoder_dir", type=str, default="/data/llm_data/FoleyCrafter2/FoleyCrafter-main/checkpoints/vocoder")
    config.add_argument("--device", type=str, default="cuda")
    return config.parse_args()

latent_dir = "/data/llm_data/FoleyCrafter2/FoleyCrafter-main/LLM_Exp/vae_latents"

#设置vae
vae_dir = "/data/llm_data/FoleyCrafter2/FoleyCrafter-main/checkpoints/auffusion/auffusion-full-no-adapter/vae"
vae_config_path = os.path.join(vae_dir, "config.json")
vae_model_path = os.path.join(vae_dir, "diffusion_pytorch_model.bin")
# **2. 手动加载 VAE**
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

vae = AutoencoderKL.from_config(vae_config_path)  # 加载 VAE 配置
vae.load_state_dict(torch.load(vae_model_path, map_location=device))  # 加载本地权重
vae.to(device).eval()
vae.enable_slicing()

# 解析参数
config = args_parse()
device = torch.device(config.device if torch.cuda.is_available() else "cpu")

# 加载 Vocoder
vocoder_config_path = "/data/llm_data/FoleyCrafter2/FoleyCrafter-main/checkpoints/"
vocoder = Generator.from_pretrained(vocoder_config_path, subfolder="vocoder").to(device)

# 创建输出文件夹
os.makedirs(config.reconstructed_wav_dir, exist_ok=True)

# 处理 WAV 文件
wav_files = sorted([f for f in os.listdir(config.wav_dir) if f.endswith(".wav")])[:10]

# 修改后的处理流程
for wav_file in tqdm(wav_files, desc="Testing Vocoder Directly"):
    wav_path = os.path.join(config.wav_dir, wav_file)
    
    # 读取音频并预处理
    waveform, sample_rate = torchaudio.load(wav_path)

    # resampler = torchaudio.transforms.Resample(orig_freq=44100, new_freq=16000)
    # waveform = resampler(waveform)

    waveform = waveform.mean(dim=0)  # 转换为单声道
    waveform = waveform / MAX_WAV_VALUE  # 归一化到[-1,1]
    
    # 应用与训练时相同的预处理
    waveform = waveform.numpy()
    waveform = librosa.util.normalize(waveform) * 0.95  # 来自get_mel_spectrogram_from_audio
    waveform = torch.FloatTensor(waveform).unsqueeze(0).to(device)

    # 使用与训练一致的Mel参数
    spectrogram = mel_spectrogram(
        waveform,
        n_fft=2048,
        num_mels=256,
        sampling_rate=16000,
        hop_size=160,
        win_size=1024,
        fmin=0,
        fmax=8000,  # 必须与训练时一致
        center=False
    )
    
    # 应用正确的动态范围压缩
    # spectrogram = dynamic_range_compression_torch(spectrogram)

    
    print(f"Vocoder Input: {spectrogram.shape}, min: {spectrogram.min().item():.2f}, max: {spectrogram.max().item():.2f}")
    with torch.no_grad():
        print(f"Input to VAE encode(): {spectrogram.shape}, min: {spectrogram.min()}, max: {spectrogram.max()}")
        # latents = vae.encode(spectrogram).latent_dist.mean
        print(spectrogram.shape)
        spectrogram2=spectrogram
        # 调整通道数（扩展到 3 通道）
        spectrogram=spectrogram.unsqueeze(1).repeat(1,3,1,1)
        # spectrogram = spectrogram.expand(1, 3, -1, -1)  # 变成 [1, 3, 256, T]
        spectrogram = (spectrogram - spectrogram.min()) / (spectrogram.max() - spectrogram.min())
        # spectrogram = spectrogram * 2 - 1  # 归一化到 [-1, 1]
        spectrogram =(spectrogram/2+0.5).clamp(0,1)
        latents = vae.encode(spectrogram).latent_dist.sample()  # 获取.sample()作为 latent 表示

        latents = latents * vae.config.scaling_factor
        import pdb
        pdb.set_trace()
    print("Latent encoding complete.")
   

    with torch.no_grad():
        latents = latents / vae.config.scaling_factor  # 重要
        # import pdb
        # pdb.set_trace()
        spectrogram = vae.decode(latents).sample
        spectrogram = (spectrogram + 1) / 2
        spectrogram = spectrogram.squeeze(0)

   
    # 送入Vocoder
    with torch.no_grad():
        # audio = vocoder.inference(spectrogram2)[0]
        audio = vocoder.inference(spectrogram)[0]
        # audio = vocoder.inference(spectrogram, lengths=160000)[0]
        
    # 保存音频
    save_path = os.path.join(config.reconstructed_wav_dir, wav_file)

    sf.write(save_path, audio, 44100)

print("✅ Direct vocoder test completed.")

**Comment by ISSpeechSEC** [comment]
Now the problem has been solved ! Thanks!~
